File: leetcode/algorithms/word_ladder/word_ladder.cc
#include <iostream>
#include <string>
#include <utility>
#include <unordered_set>
#include <queue>
using namespace std;

class Solution {
public:
    vector<string> findNeighbours(string word, unordered_set<string>& dict) {
        vector<string> neighbours;
        int n = word.size();

        for (int i = 0; i < n; ++i) {
            string s = word;
            for (char c = 'a'; c <= 'z'; ++c) {
                s[i] = c;
                if (dict.count(s) && this->areNeighbours(word, s)) {
                    neighbours.push_back(s);
                }
            }
        }

        return neighbours;
    }

    int areNeighbours(string word, string other) {
        int n = word.size();
        int changes = 0;

        for (int i = 0; i < n; ++i) {
            if (word[i] != other[i]) {
                ++changes;
            }
            if (changes > 1) {
                return false;
            }
        }

        return true;
    }

    int ladderLength(string beginWord, string endWord, unordered_set<string>& wordDict) {
        wordDict.insert(endWord);

        queue<pair<string, int>> q;
        q.push(make_pair(beginWord, 1));

        queue<pair<string, int>> rq;
        rq.push(make_pair(endWord, 1));

        unordered_set<string> visited_q;
        visited_q.insert(beginWord);

        unordered_set<string> visited_rq;
        visited_rq.insert(endWord);

        int q_level = 1;
        int rq_level = 1;

        while (!q.empty() && !rq.empty()) {
            if (q.size() < rq.size()) {
                while (!q.empty() && q.front().second == q_level) {
                    vector<string> neighbours = this->findNeighbours(q.front().first, wordDict);

                    for (auto &word: neighbours) {
                        if (visited_q.count(word) == 0) {
                            visited_q.insert(word);
                            if (visited_rq.count(word)) {
                                return q.front().second + rq.front().second;
                            }
                            q.push(make_pair(word, q_level + 1));
                        }
                    }
                    q.pop();
                }
                ++q_level;
            } else {
                while (!rq.empty() && rq.front().second == rq_level) {
                    vector<string> neighbours = this->findNeighbours(rq.front().first, wordDict);

                    for (auto &word: neighbours) {
                        if (visited_rq.count(word) == 0) {
                            visited_rq.insert(word);
                            if (visited_q.count(word)) {
                                return rq.front().second + q.front().second;
                            }
                            rq.push(make_pair(word, rq_level + 1));
                        }
                    }
                    rq.pop();
                }
                ++rq_level;
            }
        }

        return 0;
    }
};


int main(void) {
    string a = "a";
    string b = "b";
    string c = "c";

    unordered_set<string> s;

    s.insert(a);
    s.insert(b);
    s.insert(c);

    Solution sol;

    cout << sol.ladderLength(a, c, s) << endl;

    return 0;
}

File: leetcode/algorithms/word_ladder/word_ladder.py
import collections

Node = collections.namedtuple('Node', 'word, level')

class Solution(object):
    def ladderLength(self, begin_word, end_word, word_dict):
        word_dict.add(end_word)

        start_queue = Queue(begin_word)
        end_queue = Queue(end_word)

        while start_queue and end_queue:
            if len(start_queue) < len(end_queue):
                output = self.process(start_queue, end_queue, word_dict)
                if output:
                    return output
            else:
                output = self.process(end_queue, start_queue, word_dict)
                if output:
                    return output

        return 0

    def process(self, start_queue, end_queue, word_dict):
        while start_queue and start_queue[0].level == start_queue.level:
            neighbours = self.findNeighbours(start_queue[0].word, word_dict)

            for word in neighbours:
                if not start_queue.has_visited(word):
                    start_queue.visit(word)
                    if end_queue.has_visited(word):
                        return start_queue[0].level + end_queue[0].level
                    start_queue.append(Node(word, start_queue.level + 1))

            start_queue.pop()
        start_queue.level += 1

    def findNeighbours(self, word, word_dict):
        for i, char in enumerate(word):
            s = list(word)
            for j in xrange(97, 123):
                s[i] = chr(j)
                new_word = ''.join(s)
                if new_word in word_dict and self.areNeighbours(word, new_word):
                    yield new_word

    def areNeighbours(self, word, other):
        changes = 0
        for c1, c2 in zip(word, other):
            if c1 != c2:
                changes += 1
            if changes > 1:
                return False
        return changes == 1


class Queue(object):
    def __init__(self, init=None):
        self.queue = collections.deque()
        self.level = 1
        self.visited = set()

        if init:
            self.queue.append(Node(init, self.level))
            self.visited.add(init)

    def __len__(self):
        return len(self.queue)

    def __getitem__(self, key):
        return self.queue[key]

    def visit(self, word):
        self.visited.add(word)

    def has_visited(self, word):
        return word in self.visited

    def append(self, node):
        self.queue.append(node)

    def pop(self):
        self.queue.popleft()
